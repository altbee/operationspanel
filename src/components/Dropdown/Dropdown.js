import React, { Component } from "react";
import PropTypes from "prop-types";

export default class Dropdown extends Component {
  constructor(props) {
    super(props);

    this.state = {
      opened: false,
      selected: props.list[0] || ""
    };
  }

  toggleDropdown() {
    const { opened } = this.state;
    this.setState({ opened: !opened });
  }

  selectOption(option) {
    const { handleChange } = this.props;
    this.toggleDropdown();
    this.setState({ selected: option });
    handleChange(option);
  }

  render() {
    const { list, title } = this.props;
    const { opened, selected } = this.state;
    const dropdownClass = opened ? "show" : "";

    const label = title ? (
      <div className="dropdown-label btn btn-secondary w-50">{title}</div>
    ) : (
      ""
    );

    return (
      <div
        title={`Search By ${selected}`}
        className="btn-group dropdown-component"
        role="group"
      >
        {label}
        <button
          type="button"
          className="btn btn-secondary dropdown-toggle w-100 text-right"
          data-toggle="dropdown"
          aria-haspopup="true"
          aria-expanded="false"
          onClick={this.toggleDropdown.bind(this)}
        >
          {selected}
        </button>

        <div className={`dropdown-menu dropdown-menu-right ${dropdownClass}`}>
          {list.map(d => (
            <button
              className="dropdown-item"
              key={d}
              onKeyPress={() => {}}
              onClick={() => this.selectOption(d)}
              type="button"
            >
              {d}
            </button>
          ))}
        </div>
      </div>
    );
  }
}

Dropdown.propTypes = {
  list: PropTypes.instanceOf(Array).isRequired,
  title: PropTypes.string,
  handleChange: PropTypes.func.isRequired
};

Dropdown.defaultProps = {
  title: ""
};
