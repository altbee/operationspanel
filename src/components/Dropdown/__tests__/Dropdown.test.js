import React from "react";
import { configure, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import Dropdown from "../Dropdown";

configure({ adapter: new Adapter() });

describe("<Dropdown />", () => {
  const requiredProps = {
    list: ["first", "second", "third"],
    title: "Passed title",
    handleChange: () => "Something happened"
  };

  it("Renders correctly", () => {
    const rendered = shallow(<Dropdown {...requiredProps} />);
    expect(rendered.find(".dropdown-component").exists()).toBe(true);
    expect(rendered.find(".dropdown-label").text()).toEqual(
      requiredProps.title
    );
  });

  it("Renders with default props", () => {
    const trimmedProps = { ...requiredProps };
    delete trimmedProps.title;
    const rendered = shallow(<Dropdown {...trimmedProps} />);
    expect(rendered.find(".dropdown-label").exists()).toBe(false);
  });

  it("Renders the correct amount of items", () => {
    const rendered = shallow(<Dropdown {...requiredProps} />);
    const dropdownMenu = rendered.find(".dropdown-menu");
    expect(dropdownMenu.children().length).toBe(requiredProps.list.length);
  });

  it("Toggles the dropdown visibility on dropdown click", () => {
    const rendered = shallow(<Dropdown {...requiredProps} />);
    expect(rendered.state().opened).toBe(false);
    expect(rendered.find(".dropdown-menu").hasClass("show")).toBe(false);
    rendered.find(".dropdown-toggle").simulate("click");
    expect(rendered.state().opened).toBe(true);
    expect(rendered.find(".dropdown-menu").hasClass("show")).toBe(true);
  });

  it("Updates state and dropdown button text on selection", () => {
    const rendered = shallow(<Dropdown {...requiredProps} />);
    expect(rendered.state().selected).toBe(requiredProps.list[0]);
    rendered.find(".dropdown-toggle").simulate("click");
    rendered
      .find(".dropdown-menu")
      .childAt(2)
      .simulate("click");
    expect(rendered.state().selected).toBe(requiredProps.list[2]);
    expect(rendered.find(".dropdown-toggle").text()).toEqual(
      requiredProps.list[2]
    );
  });

  it("matches the snapshot", () => {
    const tree = shallow(<Dropdown {...requiredProps} />);
    expect(tree).toMatchSnapshot();
  });
});
