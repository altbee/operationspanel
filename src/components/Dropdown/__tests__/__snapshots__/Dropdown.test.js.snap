// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> matches the snapshot 1`] = `
<div
  className="btn-group dropdown-component"
  role="group"
  title="Search By first"
>
  <div
    className="dropdown-label btn btn-secondary w-50"
  >
    Passed title
  </div>
  <button
    aria-expanded="false"
    aria-haspopup="true"
    className="btn btn-secondary dropdown-toggle w-100 text-right"
    data-toggle="dropdown"
    onClick={[Function]}
    type="button"
  >
    first
  </button>
  <div
    className="dropdown-menu dropdown-menu-right "
  >
    <button
      className="dropdown-item"
      key="first"
      onClick={[Function]}
      onKeyPress={[Function]}
      type="button"
    >
      first
    </button>
    <button
      className="dropdown-item"
      key="second"
      onClick={[Function]}
      onKeyPress={[Function]}
      type="button"
    >
      second
    </button>
    <button
      className="dropdown-item"
      key="third"
      onClick={[Function]}
      onKeyPress={[Function]}
      type="button"
    >
      third
    </button>
  </div>
</div>
`;
